pkg/tfgen/installation_docs.go

```diff
@@ -245,3 +250,106 @@ func convertExample(g *Generator, code string, exampleNumber int) (string, error
 	exampleContent += chooserEnd
 	return exampleContent, nil
 }
+
+type SectionSkipper struct {
+	shouldSkipHeader shouldSkipHeaderFunc
+}
+
+var _ parser.ASTTransformer = SectionSkipper{}
+
+func (t SectionSkipper) Transform(node *ast.Document, reader text.Reader, pc parser.Context) {
+	source := reader.Source()
+	removingLevel := 0
+	currentChild := node.FirstChild()
+	// All headings are first children of the ast.Document node.
+	// We will walk over them and remove any that match the header content we do not want.
+	// Additionally, we will track the header level and remove any content that is nested under the removed header.
+	for currentChild != nil {
+		if removingLevel > 0 {
+			if child, ok := currentChild.(*ast.Heading); ok && child.Level <= removingLevel {
+				removingLevel = 0
+			} else {
+				nextNode := currentChild.NextSibling()
+				currentChild.Parent().RemoveChild(currentChild.Parent(), currentChild)
+				currentChild = nextNode
+				continue
+			}
+		}
+		// In case our level got reset, we check again
+		if removingLevel == 0 {
+			switch child := currentChild.(type) {
+			case *ast.Heading:
+				header := child.Text(source)
+				if t.shouldSkipHeader(string(header)) {
+					removingLevel = child.Level
+					nextNode := child.NextSibling()
+					child.Parent().RemoveChild(child.Parent(), child)
+					currentChild = nextNode
+					continue
+				}
+			}
+		}
+		// Move to next node in base case.
+		currentChild = currentChild.NextSibling()
+	}
+}
+
+type shouldSkipHeaderFunc = func(headerText string) bool
+
+// SkipSectionByHeaderContent takes a document's content byte array and a callback function that contains logic on
+// whether a given header should be skipped.
+// It uses "github.com/yuin/goldmark" parser in conjunction with "github.com/teekennedy/goldmark-markdown" renderer
+// to walk the markdown document tree and apply the SectionSkipper's transformer
+func SkipSectionByHeaderContent(content []byte, shouldSkipHeader shouldSkipHeaderFunc) ([]byte, error) {
```

Comment:
I don't love code comments that re-iterate how a function is implemented or what is in the function signature.

If you want to include general description of how a function works, I would move that to below the signature, since it is for people who are editing the function, not for function callers.

```suggestion
// SkipSectionByHeaderContent removes headers where shouldSkipHeader(header) returns true, along with any text under the header.
//
// content is assumed to be Github flavored markdown when parsing.
//
// shouldSkipHeader is called on the raw header text, like this:
//
//    shouldSkipHeader("My Header")
//
// *not* like this:
//
//    // This is wrong
//    shouldSkipHeader("## My Header\n")
func SkipSectionByHeaderContent(content []byte, shouldSkipHeader shouldSkipHeaderFunc) ([]byte, error) {
	// <I don't think this is necessary, since the code already says this>
	// It uses "github.com/yuin/goldmark" parser in conjunction with "github.com/teekennedy/goldmark-markdown" renderer
	// to walk the markdown document tree and apply the SectionSkipper's transformer
```

Reply:
I generally find callback functions hard to read, and since I do expect this function to be used externally as part of a DocsEdit rule, it would be helpful to include information about the callback specifics. Naming only goes so far.

Reply:
We can make it more obvious that `shouldSkipHeaderFunc` is a callback by inlining the private type:

```patch
-func SkipSectionByHeaderContent(content []byte, shouldSkipHeader shouldSkipHeaderFunc) ([]byte, error) {
+func SkipSectionByHeaderContent(
+    content []byte, 
+    shouldSkipHeader func(header string) (shouldSkip bool),
+) ([]byte, error) {
```

We should probably do this anyway to avoid exposing a public function that takes a private type (and to minamise the pollution of the `tfgen` namespace.

---

I'm happy to include info on how the callback is used. My suggestion included:

```patch
+// content is assumed to be Github flavored markdown when parsing.
+//
+// shouldSkipHeader is called on the raw header text, like this:
+//
+//    shouldSkipHeader("My Header")
+//
+// *not* like this:
+//
+//    // This is wrong
+//    shouldSkipHeader("## My Header\n")
```

Reply:
Does this look better to you?